panel/src/components/Forms/FormButtons.vue

```diff
@@ -22,29 +58,174 @@
+      // heartbeat listener again
+      if (this.hasChanges === false) {
+        this.onHeartbeat(this.listenToLock);
+        console.log('start listening...');
+      }
     }
   },
   created() {
     this.$events.$on("keydown.cmd.s", this.save);
+    setInterval(() => {
+      this.now = dayjs();
+    }, 1000);
   },
   destroyed() {
     this.$events.$off("keydown.cmd.s", this.save);
   },
   methods: {
-    reset() {
+    listenToLock() {
+      // call API to check for lock
+      return this.$api.get(this.$route.path + "/lock").then(response => {
+
+        // if a lock is set and it's not from the current user
+        // dispatch lock information to store
+        if (response.lock === true && response.user !== this.user) {
+          this.$store.dispatch("form/lock", {
+            user: response.user,
+            email: response.email,
+            time: parseInt(response.time)
+          });
+          console.log("@listenToLock: hasLock");
+
+
+        } else {
+          // if not, trigger reload of model content if the lock
+          // wasn't from the current user and remove lock
+          // information from store
+          if (this.hasLock && this.lock.user !== this.user) {
+            this.$events.$emit("model.reload");
+            console.log("$emit(model.reload)");
+          }
+
+          this.$store.dispatch("form/lock", null);
+          console.log("@listenToLock: no lock");
+        }
+      });
+    },
+    setLock() {
+      console.log("@setLock");
+      return this.$api.patch(this.$route.path + "/lock", { user: this.user });
+    },
+    unsetLock() {
+      console.log("@unsetLock");
+      return this.$api.delete(this.$route.path + "/lock").then(() => {
+        // when lock has been remove in backend,
+        // remove it from store and start listening
+        // to other locks again
+        this.$store.dispatch("form/lock", null);
+        this.onHeartbeat(this.listenToLock);
+        console.log('lock was unset => listen again...');
+      });
+    },
+    setUnlock(user) {
+      console.log("@setUnlock");
+      return this.$api.patch(this.$route.path + "/unlock", {
+        user: user
+      });
+    },
+    unsetUnlock() {
+      console.log("@unsetUnlock");
+      return this.$api.delete(this.$route.path + "/unlock", {
+        user: this.user
+      }).then(() => {
+        // when unlock has been remove in backend,
+        // remove it from store and listen
+        // to locks
+        this.$store.dispatch("form/unlock", null);
+        this.onHeartbeat(this.listenToLock);
+        console.log('unlock was unset => listen again...');
+      })
+    },
+    onDiscard() {
       this.$store.dispatch("form/revert", this.id);
+      this.unsetUnlock();
     },
-    save(e) {
+    onDownload() {
+      let content = "";
+
+      Object.keys(this.unlock).forEach(key => {
+        content += key + ": \n" + this.unlock[key];
+        content += "\n\n----\n\n";
+      });
```

Comment:
The backend implementation in `Data\Txt` has a few special cases that are not covered here. But I don't think we should duplicate the logic in the frontend either. Two options:

- Ignore the special cases, which means that the downloaded file might not be usable 1:1, but is a well-enough backup.
- Create a helper API route that can be passed the `this.unlock` object and that will then throw the object at `Data\Txt::encode()` and return the resulting string. Advantage: That helper route could actually already trigger the download to make the link hack obsolete.

Reply:
The download file isn’t usable 1:1 anyways at the moment since it only includes the changes.

Reply:
That's right. Maybe we can keep the basic implementation for now and extend it later if there's demand for it.